kratos/utilities/normal_calculation_utils.cpp

```diff
@@ -69,99 +289,106 @@ void NormalCalculationUtils::CalculateOnSimplex(
-            for(auto& r_node: r_cond.GetGeometry()) {
-                r_node.FastGetSolutionStepValue(NORMAL) = zero;
-            }
-        }
+    for(auto& r_cond : rModelPart.Conditions()) {
+        GeometryType& r_geometry = r_cond.GetGeometry();
+        Node<3>::Pointer paux = r_geometry(0);
+        r_geometry(0) = r_geometry(1);
+        r_geometry(1) = paux;
     }
 
-    this->CalculateOnSimplex(rModelPart.Conditions(),Dimension);
-    rModelPart.GetCommunicator().AssembleCurrentData(NORMAL);
+    KRATOS_CATCH("")
 }
 
 /***********************************************************************************/
 /***********************************************************************************/
 
-void NormalCalculationUtils::SwapNormals(ModelPart& rModelPart)
+void NormalCalculationUtils::ComputeUnitNormalsFromAreaNormals(ModelPart& rModelPart)
 {
-    KRATOS_TRY
+    // We iterate over nodes
+    auto& r_nodes_array = rModelPart.Nodes();
+    const auto it_node_begin = r_nodes_array.begin();
+    const int num_nodes = static_cast<int>(r_nodes_array.size());
 
-    for(ModelPart::ConditionsContainerType::iterator it=rModelPart.ConditionsBegin(); it!=rModelPart.ConditionsEnd(); it++) {
-        Node<3>::Pointer paux = it->GetGeometry()(0);
-        it->GetGeometry()(0) = it->GetGeometry()(1);
-        it->GetGeometry()(1) = paux;
-        //it->GetGeometry()(0).swap(it->GetGeometry()(1));
+    #pragma omp parallel for
+    for (int i = 0; i < num_nodes; ++i) {
+        auto it_node = it_node_begin + i;
+
+        array_1d<double, 3>& r_normal = it_node->FastGetSolutionStepValue(NORMAL);
+        const double norm_normal = norm_2(r_normal);
+
+        if (norm_normal > std::numeric_limits<double>::epsilon()) r_normal /= norm_normal;
+        else KRATOS_ERROR_IF(it_node->Is(INTERFACE)) << "ERROR:: ZERO NORM NORMAL IN NODE: " << it_node->Id() << std::endl;
     }
 
-    KRATOS_CATCH("")
+    // For MPI: correct values on partition boundaries
+    rModelPart.GetCommunicator().AssembleCurrentData(NORMAL);
 }
 
 /***********************************************************************************/
 /***********************************************************************************/
 
 void NormalCalculationUtils::CalculateNormal2D(
-    ConditionsArrayType::iterator it,
-    array_1d<double,3>& An
+    ConditionsArrayType::iterator itCond,
+    array_1d<double,3>& rAn
     )
 {
-    Geometry<Node<3> >& pGeometry = (it)->GetGeometry();
+    const GeometryType& r_geometry = itCond->GetGeometry();
 
-    An[0] =    pGeometry[1].Y() - pGeometry[0].Y();
-    An[1] = - (pGeometry[1].X() - pGeometry[0].X());
-    An[2] =    0.00;
+    rAn[0] =    r_geometry[1].Y() - r_geometry[0].Y();
+    rAn[1] = - (r_geometry[1].X() - r_geometry[0].X());
+    rAn[2] =    0.00;
 
-    array_1d<double,3>& normal = (it)->GetValue(NORMAL);
-    noalias(normal) = An;
+    array_1d<double,3>& r_normal = itCond->GetValue(NORMAL);
+    noalias(r_normal) = rAn;
```

Comment:
```suggestion
    itCond->SetValue(NORMAL, rAn);
```
This will also work ryt?, it will be helpfull in OMP parallelism. :)